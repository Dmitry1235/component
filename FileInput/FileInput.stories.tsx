import React, { useRef, useCallback } from 'react';
import { action } from '@storybook/addon-actions';
import { FileInput } from './FileInput';
import { FileInputDocs } from './docs/FileInput.docs';

import { withKnobs, array, number } from '@storybook/addon-knobs';

export default {
  title: 'Base/FileInput',
  decorators: [withKnobs],
  parameters: {
    docs: {
      page: FileInputDocs,
    },
  },
};

const TestFileInput = () => {
  const inputRef = useRef<HTMLInputElement>(null);

  // Служит для открытия меню выдора файла по нажатию на Label
  const handleClickLabel = useCallback(() => {
    if (inputRef && inputRef.current) {
      inputRef.current.click();
    }
  }, [inputRef]);

  return (
    <div
      style={{
        position: 'fixed',
        background: 'red',
        minHeight: '400px',
        display: 'flex',
        flexDirection: 'column',
      }}
    >
      <div style={{ height: '50px', background: 'green' }}></div>
      <FileInput
        permittedFormats={array('permittedFormats', ['png', 'pdf', 'jpeg', 'doc', 'docx'])}
        maxSize={number('maxSize', 5 * 1000 * 1000)}
        ref={inputRef}
        onChange={(file, error) => {
          action('onChange')(file, error);
        }}
      >
        <div>Перетащите сюда файлы или</div>
        <label onClick={handleClickLabel}>выберите из списка</label>
      </FileInput>
    </div>
  );
};

export const example = () => <TestFileInput />;
