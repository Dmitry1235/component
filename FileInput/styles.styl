.FileInput {
  display flex
  height 100%
  flex 1 1 auto

  &_move > &__label {
    position absolute
    top 0
    left 0
    bottom 0
    z-index $zTopLay
    border 2px dashed $blueBcColor
    border-radius 15px
  }

  &__label {
    display flex
    width 100%
    min-width max-content
    padding 0 10px
    flex-direction column
    font-size 17px
    line-height 24px
    justify-content center
    align-items center

    @media $phone {
      min-width inherit
      padding 0
      justify-content flex-start
    }
  }

  &_move {
    & ^[0]__label {
      background $activeTextColor
    }

    & ^[0]__mask {
      width 100%
      height 100%
    }
  }

  &__mask {
    position absolute
    top 0
    right 0
    left 0
    bottom 0
    z-index 100
    width 0
    height 0
  }

  &__input {
    position absolute
    width 0
  }
}