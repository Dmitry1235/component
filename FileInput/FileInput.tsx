import React, { useState, useCallback, useRef, useMemo } from 'react';
import cn from 'classnames/bind';

//Libs
import { readFile } from 'libs/documents';
// Styles
import styles from './styles.styl';
const cx = cn.bind(styles);

export type FileInputType = {
  lastModified: number;
  lastModifiedDate: Date;
  name: string;
  size: number;
  type: string;
};

export type ErrorFileInputType = {
  file: FileInputType;
  isMaxSize: boolean;
  isPermittedFormat: boolean;
  isEmptySize: boolean;
};

export interface FileInputProps {
  className?: string;
  multiple?: boolean;
  permittedFormats?: string[];
  maxSize?: number;
  onChange: (files: FileInputType[], filesError: ErrorFileInputType[]) => void;
  children: React.ReactNode;
}

const ACCEST_TYPE_FILES = {
  'application/pdf': 'pdf',
  'application/x-pdf': 'x-pdf',
  'application/msword': 'doc',
  'application/vnd.openxmlformats-officedocument.wordprocessingml.document': 'docx',
  'image/jpeg': 'jpeg',
  'image/pjpeg': 'pjpeg',
  'image/png': 'png',
  'image/x-png': 'x-png',
};

export const FileInput = React.forwardRef<HTMLInputElement, FileInputProps>(
  ({ className, multiple = true, permittedFormats, maxSize, onChange, children }, forwardRef) => {
    const contentRef = useRef<HTMLDivElement>(null);
    const [isMoveFile, setMoveFile] = useState(false);

    const heightContainer = useMemo(() => {
      return contentRef.current && contentRef.current.clientHeight;
    }, [contentRef.current]);

    // Проверка файла на допустимые форматы
    const isPermittedFormat = useCallback(
      (file) => {
        if (permittedFormats) {
          const selectFile = file || '';
          const typeFile = selectFile ? ACCEST_TYPE_FILES[selectFile.type] : '';
          return permittedFormats.includes(typeFile);
        }
        return true;
      },
      [permittedFormats],
    );

    // Проверка файла на допустимые форматы //bate
    const isMaxSize = useCallback(
      (file) => {
        if (maxSize) {
          return file.size <= maxSize;
        }
        return true;
      },
      [maxSize],
    );

    // Проверка файла на валидность
    const handleChange = useCallback(
      (file) => {
        const files = file || [];
        const errorFiles: any = [];

        const promiseSucces = [];

        for (let i = 0; i < files.length; i++) {
          const isEmptySize = files[i].size === 0;
          const isMaxSizeError = isMaxSize(files[i]);
          const isPermittedFormatError = isPermittedFormat(files[i]);

          if (isMaxSizeError && isPermittedFormatError && !isEmptySize) {
            promiseSucces.push(readFile(files[i]));
          } else {
            errorFiles.push({
              file: files[i],
              isMaxSize: !isMaxSizeError,
              isPermittedFormat: !isPermittedFormatError,
              isEmptySize: isEmptySize,
            });
          }
        }

        Promise.all(promiseSucces).then((res) => {
          const succesFiles = res.map((item: any) => {
            Object.defineProperty(item.file, 'src', {
              writable: true,
              value: item.src,
            });

            return item.file;
          });

          onChange(succesFiles, errorFiles);
          setMoveFile(false);
        });
      },
      [onChange, isMaxSize, isPermittedFormat],
    );

    // Срабатывает при выборе файла из меню
    const onChangeInput = useCallback(
      (e: React.ChangeEvent<HTMLInputElement>) => {
        const file = e.target.files;
        handleChange(file);
        e.preventDefault();
        return false;
      },
      [handleChange],
    );

    // Срабатывает когда пользователь "сбрасывает" перетаскиваемый элемент
    const handleDrop = useCallback(
      (e: React.DragEvent<HTMLDivElement>) => {
        const file = e.dataTransfer.files;
        handleChange(file);
        e.preventDefault();
        e.stopPropagation();
      },
      [handleChange],
    );

    // Выполняется когда элемент перемещают над допустимой зоной
    const handleDragOver = useCallback(
      (e: React.DragEvent<HTMLDivElement>) => {
        e.preventDefault();
        e.stopPropagation();

        if (!isMoveFile) {
          setMoveFile(true);
        }

        return false;
      },
      [isMoveFile],
    );

    // Выполняется когда элемент выходит из допустимой зоны для переноса
    const handleDragLeave = useCallback((e: React.DragEvent<HTMLDivElement>) => {
      e.preventDefault();
      e.stopPropagation();

      setMoveFile(false);
      return false;
    }, []);

    // Получение строки с разрешенными форматами файлов
    const acceptFormats = useMemo(() => {
      if (permittedFormats) {
        return `.${permittedFormats.join(', .')}`;
      }
      return '';
    }, [permittedFormats]);

    return (
      <div
        className={cx(
          'FileInput',
          { FileInput_move: isMoveFile },
          { [`${className}_move`]: className && isMoveFile },
          className,
        )}
        style={isMoveFile ? { minHeight: `${heightContainer || 0}px` } : {}}
      >
        <div ref={contentRef} className={cx('FileInput__label')} onDragOver={handleDragOver}>
          {children}
          <div
            className={cx('FileInput__mask')}
            onDragLeave={handleDragLeave}
            onDrop={handleDrop}
          />
        </div>
        <input
          className={cx('FileInput__input')}
          type="file"
          onChange={onChangeInput}
          multiple={multiple}
          ref={forwardRef}
          accept={acceptFormats}
        />
      </div>
    );
  },
);

FileInput.displayName = 'FileInput';
