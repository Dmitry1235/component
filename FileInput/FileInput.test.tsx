import React from 'react';
import { FileInput } from './FileInput';
import renderer from 'react-test-renderer';

describe('FileInput:: Snapshots', () => {
  it('Default', () => {
    const component = renderer.create(<FileInput onChange={() => {}}>Загрузите файл</FileInput>);
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
