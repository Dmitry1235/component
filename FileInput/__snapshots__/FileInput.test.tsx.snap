// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileInput:: Snapshots Default 1`] = `
<div
  className="FileInput"
  style={Object {}}
>
  <div
    className="FileInput__label"
    onDragOver={[Function]}
  >
    Загрузите файл
    <div
      className="FileInput__mask"
      onDragLeave={[Function]}
      onDrop={[Function]}
    />
  </div>
  <input
    accept=""
    className="FileInput__input"
    multiple={true}
    onChange={[Function]}
    type="file"
  />
</div>
`;
